text/0000-localizing-api-messages.md

```diff
@@ -0,0 +1,135 @@
+
+- Start Date: 03-29-2022
+- RFC PR: 
+- FOLIO Issue: 
+- Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
+- Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
+- Sub Group :
+
+# Localizing API (Backend) Messages
+
+## Summary
+  The purpose of this RFC is to define an approach that allows API developer to provide
+messages in the API response depending on the language preference set in the request. 
+
+
+## Motivation
+
+- Help users understand the information coming from the application backend using a language they understand better
+- To allow FOLIO platform to be used by users across the world 
+- To create a tremendous opportunity for growth that we can never could have achieved within just one country.
+
+### Terminology
+- Static Message - E.g. _This is something you need translate_
+- Static Message with placeholder - _You can ONLY renew {n} times_
+ 
+
+### In Scope Requirements/Use cases
+- Return localized messages based on the value passed in the accept-language header
+- Handle static messages, static messages with placeholders
+
+### Out of Scope Requirements/Use cases
+- [Plural Syntax](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Pluralsyntax)
+- Naming convention to be used for keys
+- Process for managing (decoupled from lokalise.com) translations
+- Returning formatted (HTML/Markdown) messages
+- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords)to break messages
+- Support customization per tenant
+- Process for back porting API/Backend messages similar to what we have for the front end 
+- Support for controlled vocabulary (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)
+
+## Detailed Explanation/Design
+- Locale specific translations MUST be stored in a properties file
+- Properties file naming convention
+  - default locale (en_US) - somename.properties
+  - specific locale - somename_fr.properties (For french),  somename_de.properties (For german)
+- Properties file MUST BE loaded using the classpath. AVOID using file locations
```

Comment:
I wanted to standardize on the mechanism that was used load these files without adding additional complexity. Had I suggested that we use file locations, I need to specify where these will be located. Absolute vs relative path. If it is absolute, how will I pass on this information to the code that needs it. System properties, env var etc.. If it is relative, relative to what. Wanted to avoid all these complexities :)

Reply:
> I wanted to standardize on the mechanism that was used load these files without adding additional complexity. 

I understand not wanting to add unnecessary complexity.

> Had I suggested that we use file locations, I need to specify where these will be located. Absolute vs relative path. If it is absolute, how will I pass on this information to the code that needs it. System properties, env var etc.. If it is relative, relative to what. Wanted to avoid all these complexities :)

I agree that for Java loading via the class path is better than file paths.

I think that we need to acknowledge that this has trade offs in complexity in other ways:
* Folks need to understand the guidance is Java (or at least JVM specific)
* Folks need to understand that the location of these files might vary between modules
* The operational processes for managing translations need to accommodate these variations

Reply:
How properties get loaded is an implementation detail that should be out of scope of an RFC. If there are best practices they can be published on dev.folio.org or wiki.folio.org.

The translation workflow might be out of scope. However, each translation workflow need to pick up some files from the GitHub repository and need to store translated files in the GitHub repository. Defining a standard location will help the translation workflow devops and the translators. Therefore this interface should be in scope and been standardized.

Reply:
I think we agreed to recommend that language/framework specific best practices MUST be followed when loading translation files at runtime